WooCommerce/WooCommerceTests/Tools/InfiniteScroll/PaginationTrackerTests.swift

```diff
@@ -0,0 +1,120 @@
+            syncedPages.append(PageInformation(pageNumber: pageNumber, pageSize: pageSize, reason: reason))
+            onCompletion?(.success(false))
+        }
+
+        // Action
+        paginationTracker.syncFirstPage()
+        paginationTracker.ensureNextPageIsSynced()
+
+        // Assert
+        XCTAssertEqual(syncedPages, [
+            PageInformation(pageNumber: 0, pageSize: 30, reason: nil)
+        ])
+    }
+
+    func test_resyncing_after_syncing_first_page_triggers_delegate_to_sync_the_first_page_again() {
+        // Arrange
+        var syncedPages = [PageInformation]()
+        mockDelegate.onSync = { pageNumber, pageSize, reason, onCompletion in
+            syncedPages.append(PageInformation(pageNumber: pageNumber, pageSize: pageSize, reason: reason))
+            onCompletion?(.success(true))
+        }
+
+        // Action
+        paginationTracker.syncFirstPage()
+        paginationTracker.resync(reason: "Testing", onCompletion: nil)
+
+        // Assert
+        XCTAssertEqual(syncedPages, [
+            PageInformation(pageNumber: 0, pageSize: 30, reason: nil),
+            PageInformation(pageNumber: 0, pageSize: 30, reason: "Testing")
+        ])
+    }
+
+    func test_ensuring_next_page_without_syncing_first_page_triggers_delegate_to_sync_the_first_page() {
+        // Arrange
+        var syncedPages = [PageInformation]()
+        mockDelegate.onSync = { pageNumber, pageSize, reason, onCompletion in
+            syncedPages.append(PageInformation(pageNumber: pageNumber, pageSize: pageSize, reason: reason))
+            onCompletion?(.success(true))
+        }
+
+        // Action
+        paginationTracker.ensureNextPageIsSynced()
+
+        // Assert
+        XCTAssertEqual(syncedPages, [PageInformation(pageNumber: 0, pageSize: 30, reason: nil)])
+    }
+
+    func test_ensuring_next_page_after_syncing_first_page_with_failure_triggers_delegate_to_sync_the_first_page_twice() {
+        // Arrange
+        var syncedPages = [PageInformation]()
+        mockDelegate.onSync = { pageNumber, pageSize, reason, onCompletion in
+            syncedPages.append(PageInformation(pageNumber: pageNumber, pageSize: pageSize, reason: reason))
+            onCompletion?(.failure(NSError(domain: "", code: 1, userInfo: nil)))
+        }
+
+        // Action
+        paginationTracker.syncFirstPage()
+        paginationTracker.ensureNextPageIsSynced()
+
+        // Assert
+        XCTAssertEqual(syncedPages, [
+            PageInformation(pageNumber: 0, pageSize: 30, reason: nil),
+            PageInformation(pageNumber: 0, pageSize: 30, reason: nil)
+        ])
+    }
+}
+
+private struct PageInformation: Equatable {
+    let pageNumber: Int
+    let pageSize: Int
+    let reason: String?
+}
+
+/// `PaginationTrackerDelegate` Closure-based wrapper, for unit testing purposes.
+///
+private class MockPaginationTrackerDelegate: PaginationTrackerDelegate {
+    typealias OnSyncClosure = (Int, Int, String?, SyncCompletion?) -> Void
```

Comment:
Based on the way we used the values in the tests, it looks like we can save some words to read by using the `PageInformation` instead:

```suggestion
    typealias OnSyncClosure = (PageInformation, SyncCompletion?) -> Void
```

Reply:
Great suggestion! looks much more concise now: https://github.com/woocommerce/woocommerce-ios/pull/2750/commits/83045affdd260821e25cadfc58093307ee21e378